@amaltaro using the name of the file that you provied instead of `dqm.root` worked indeed! i will push a new commit.

I paste below some logging of the successful request which may be useful for future-us. Notice that I am artificially making it fail just to retrieve the logs (with some debugging information) in case of a successfull request. 

### py3 log

```plaintext
#1 testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t) ... ERROR

======================================================================
ERROR: testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py", line 62, in testA_Upload
    raise Exception
Exception: 
-------------------- >> begin captured logging << --------------------
root: INFO: Steps.Executor logging started
root: INFO: Using proxy file: /home/dmwm/certs-mount/x509up_u99307
root: INFO: Using CA certificate path: None
root: INFO: HTTP POST upload arguments:
  ==> checksum: md5:776e66d494a6eeabd52b1d7fb50e9bd4
  ==> size: 150553620

WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: Found 142 default trusted CA certificates.
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: SSL context manager created with the following settings:
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   check_hostname : True
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   options : Options.OP_ALL|OP_NO_SSLv3|OP_NO_SSLv2|OP_CIPHER_SERVER_PREFERENCE|OP_SINGLE_DH_USE|OP_SINGLE_ECDH_USE|OP_NO_COMPRESSION
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   protocol : _SSLMethod.PROTOCOL_TLS
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_flags : VerifyFlags.VERIFY_X509_TRUSTED_FIRST
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_mode : VerifyMode.CERT_REQUIRED
root: DEBUG: encode value - <class 'str'>, md5:776e66d494a6eeabd52b1d7fb50e9bd4
root: DEBUG: encode payload - <class 'bytes'>, b'md5:776e66d494a6eeabd52b1d7fb50e9bd4'
root: DEBUG: encode value - <class 'int'>, 150553620
root: DEBUG: encode payload - <class 'bytes'>, b'150553620'
root: DEBUG: encode body (without binary file) -<class 'bytes'>, b'------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="checksum"\r\n\r\nmd5:776e66d494a6eeabd52b1d7fb50e9bd4\r\n------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="size"\r\n\r\n150553620\r\n------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="file"; filename="DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root"\r\nContent-Type: application/octet-stream\r\nContent-Length: 150553620\r\n'
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG: POST method to https://cmsweb-testbed.cern.ch/dqm/dev//data/put
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG:   with the following headers: {'Accept-encoding': 'gzip', 'User-agent': 'WMAgent python/3.8.2', b'Content-type': b'multipart/form-data; boundary=----------=_DQM_FILE_BOUNDARY_=-----------'}
root: DEBUG: headers: <class 'http.client.HTTPMessage'> Date: Fri, 13 Aug 2021 11:03:55 GMT
Server: CherryPy/5.4.0
Set-Cookie: cms-auth=a1db9d8e44eac6a056b1b6d5b0d937561bd90cef6b250f9c42967d03247917ceb85188e2ef081d14;path=/;secure;httponly;expires=Thu, 01-Jan-1970 00:00:01 GMT
Dqm-Status-Detail: Wrote 150553620 bytes
Content-Length: 376
Dqm-Status-Code: 100
Dqm-Status-Message: File saved
Content-Type: text/html;charset=utf-8
CMS-Server-Time: D=6784126 t=1628852635676700
Connection: close


root: DEBUG: data: <class 'bytes'> b'Thanks.\nThe server received your file. However this does not mean that the file was indexed successfully.\nA different server process will now check the filename to determine in which folder to archive it.\nAfter that, another process will attempt to index it. This might take some minutes.\nIn case of doubt, please check the server logs of the receive-daemon and import-daemon.'
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
XML: /home/dmwm/wmcore_unittest/WMCore/nosetests.xml
----------------------------------------------------------------------
Ran 1 test in 9.509s

FAILED (errors=1)
Testing complete, there are now 1 threads
```

### py2 log

```plaintext
#1 testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t) ... ERROR

======================================================================
ERROR: testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py", line 62, in testA_Upload
    raise Exception
Exception: 
-------------------- >> begin captured logging << --------------------
root: INFO: Steps.Executor logging started
root: INFO: Using proxy file: /home/dmwm/certs-mount/x509up_u99307
root: INFO: Using CA certificate path: None
root: INFO: HTTP POST upload arguments:
  ==> checksum: md5:776e66d494a6eeabd52b1d7fb50e9bd4
  ==> size: 150553620

WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: Found 142 default trusted CA certificates.
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: SSL context manager created with the following settings:
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   check_hostname : True
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   options : 2197947391
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   protocol : 2
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_flags : 32768
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_mode : 2
root: DEBUG: encode value - <type 'str'>, md5:776e66d494a6eeabd52b1d7fb50e9bd4
root: DEBUG: encode payload - <type 'str'>, md5:776e66d494a6eeabd52b1d7fb50e9bd4
root: DEBUG: encode value - <type 'int'>, 150553620
root: DEBUG: encode payload - <type 'str'>, 150553620
root: DEBUG: encode body (without binary file) -<type 'str'>, ------------=_DQM_FILE_BOUNDARY_=-----------
Content-Disposition: form-data; name="checksum"

md5:776e66d494a6eeabd52b1d7fb50e9bd4
------------=_DQM_FILE_BOUNDARY_=-----------
Content-Disposition: form-data; name="size"

150553620
------------=_DQM_FILE_BOUNDARY_=-----------
Content-Disposition: form-data; name="file"; filename="DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root"
Content-Type: application/octet-stream
Content-Length: 150553620

WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG: POST method to https://cmsweb-testbed.cern.ch/dqm/dev//data/put
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG:   with the following headers: {'Content-length': '150554231', 'Content-type': 'multipart/form-data; boundary=----------=_DQM_FILE_BOUNDARY_=-----------', 'Accept-encoding': 'gzip', 'User-agent': 'WMAgent python/2.7.13'}
root: DEBUG: headers: <type 'instance'> Date: Fri, 13 Aug 2021 11:03:57 GMT
Server: CherryPy/5.4.0
Set-Cookie: cms-auth=c6bbcbb76af633e7876c40bb2a02b99053c8353f76dd7bcc2c9d133f68bb4236ff48ee3796edcf3f;path=/;secure;httponly;expires=Thu, 01-Jan-1970 00:00:01 GMT
Dqm-Status-Detail: Wrote 150553620 bytes
Content-Length: 376
Dqm-Status-Code: 100
Dqm-Status-Message: File saved
Content-Type: text/html;charset=utf-8
CMS-Server-Time: D=6561068 t=1628852637711378
Connection: close

root: DEBUG: data: <type 'str'> Thanks.
The server received your file. However this does not mean that the file was indexed successfully.
A different server process will now check the filename to determine in which folder to archive it.
After that, another process will attempt to index it. This might take some minutes.
In case of doubt, please check the server logs of the receive-daemon and import-daemon.
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
XML: /home/dmwm/wmcore_unittest/WMCore/nosetests.xml
----------------------------------------------------------------------
Ran 1 test in 10.175s

FAILED (errors=1)
Testing complete, there are now 1 threads

```

--- a/src/python/WMCore/Services/HTTPS/HTTPSAuthHandler.py
+++ b/src/python/WMCore/Services/HTTPS/HTTPSAuthHandler.py
@@ -5,13 +5,24 @@ See usage example in:
 src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py
 """
 from __future__ import print_function, division
+
 import logging
-import urllib2
-import httplib
 import ssl
+try:
+    # python2
+    import urllib2
+    import httplib
+    HTTPSHandler = urllib2.HTTPSHandler
+    HTTPSConnection = httplib.HTTPSConnection
+except:
+    # python3
+    import urllib.request
+    import http.client
+    HTTPSHandler = urllib.request.HTTPSHandler
+    HTTPSConnection = http.client.HTTPSConnection
 
 
-class HTTPSAuthHandler(urllib2.HTTPSHandler):
+class HTTPSAuthHandler(HTTPSHandler):
     """
     HTTPS authentication class to provide a ssl context with the certificates.
     """
@@ -36,15 +47,15 @@ class HTTPSAuthHandler(urllib2.HTTPSHandler):
             self.logger.info("  protocol : %s", self.ctx.protocol)  # default to 2 (PROTOCOL_SSLv23)
             self.logger.info("  verify_flags : %s", self.ctx.verify_flags)  # default to 0 (VERIFY_DEFAULT)
             self.logger.info("  verify_mode : %s", self.ctx.verify_mode)  # default to 2 (CERT_REQUIRED)
-            urllib2.HTTPSHandler.__init__(self, debuglevel=level, context=self.ctx)
+            HTTPSHandler.__init__(self, debuglevel=level, context=self.ctx)
         else:
             self.logger.info("Certificate not provided for HTTPSHandler")
-            urllib2.HTTPSHandler.__init__(self, debuglevel=level)
+            HTTPSHandler.__init__(self, debuglevel=level)
 
     def get_connection(self, host, **kwargs):
         if self.ctx:
-            return httplib.HTTPSConnection(host, context=self.ctx, **kwargs)
-        return httplib.HTTPSConnection(host)
+            return HTTPSConnection(host, context=self.ctx, **kwargs)
+        return HTTPSConnection(host)
 
     def https_open(self, req):
         """

--- a/src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py
+++ b/src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py
@@ -12,12 +12,29 @@ from future.utils import viewitems
 import logging
 import os
 import sys
-import urllib2
 from io import BytesIO
 from functools import reduce
 from gzip import GzipFile
 from hashlib import md5
 from mimetypes import guess_type
+
+try:
+    # python2
+    import urllib2
+    HTTPError = urllib2.HTTPError
+    OpenerDirector = urllib2.OpenerDirector
+    Request = urllib2.Request
+    ProxyHandler = urllib2.ProxyHandler
+except:
+    # python3
+    import urllib.request
+    HTTPError = urllib.error.HTTPError
+    OpenerDirector = urllib.request.OpenerDirector
+    Request = urllib.request.Request
+    ProxyHandler = urllib.request.ProxyHandler
+
+from Utils.Utilities import decodeBytesToUnicode, encodeUnicodeToBytesConditional, encodeUnicodeToBytes
+from Utils.PythonVersion import PY3, PY2
 
 from WMCore.FwkJobReport.Report import Report
 from WMCore.Services.HTTPS.HTTPSAuthHandler import HTTPSAuthHandler
@@ -138,7 +155,7 @@ class DQMUpload(Executor):
             return m
 
         with open(filename, 'rb') as fd:
-            contents = iter(lambda: fd.read(blockSize), '')
+            contents = iter(lambda: fd.read(blockSize), b'')
             m = reduce(upd, contents, md5())
 
         args['checksum'] = 'md5:%s' % m.hexdigest()
@@ -164,7 +181,7 @@ class DQMUpload(Executor):
                 else:
                     msg = 'HTTP upload finished succesfully with response:\n' + msg
                     logging.info(msg)
-        except urllib2.HTTPError as ex:
+        except HTTPError as ex:
             msg = 'HTTP upload failed with response:\n'
             msg += '  Status code: %s\n' % ex.hdrs.get("Dqm-Status-Code", None)
             msg += '  Message: %s\n' % ex.hdrs.get("Dqm-Status-Message", None)
@@ -191,23 +208,35 @@ class DQMUpload(Executor):
         multi-part/form-data. We don't actually need to know what we are
         uploading here, so just claim it's all text/plain.
         """
-        boundary = '----------=_DQM_FILE_BOUNDARY_=-----------'
-        (body, crlf) = ('', '\r\n')
+        boundary = b'----------=_DQM_FILE_BOUNDARY_=-----------'
+        (body, crlf) = (b'', b'\r\n')
         for (key, value) in viewitems(args):
-            payload = str(value)
-            body += '--' + boundary + crlf
-            body += ('Content-Disposition: form-data; name="%s"' % key) + crlf
+            logging.debug("encode value - %s, %s", type(value), value)
+            if PY2:
+                payload = str(value)
+            elif PY3:
+                payload = value
+                if not isinstance(payload, bytes):
+                    payload = str(payload)
+                payload = encodeUnicodeToBytes(payload)
+                key = encodeUnicodeToBytes(key)
+            logging.debug("encode payload - %s, %s", type(payload), payload)
+            body += b'--' + boundary + crlf
+            body += (b'Content-Disposition: form-data; name="%s"' % key) + crlf
             body += crlf + payload + crlf
         for (key, filename) in viewitems(files):
-            body += '--' + boundary + crlf
-            body += ('Content-Disposition: form-data; name="%s"; filename="%s"'
-                     % (key, os.path.basename(filename))) + crlf
-            body += ('Content-Type: %s' % self.filetype(filename)) + crlf
-            body += ('Content-Length: %d' % os.path.getsize(filename)) + crlf
-            with open(filename, "r") as fd:
+            body += b'--' + boundary + crlf
+            key = encodeUnicodeToBytesConditional(key, condition=PY3)
+            filepath = encodeUnicodeToBytesConditional(os.path.basename(filename), condition=PY3)
+            body += (b'Content-Disposition: form-data; name="%s"; filename="%s"'
+                     % (key, filepath)) + crlf
+            body += (b'Content-Type: %s' % encodeUnicodeToBytes(self.filetype(filename))) + crlf
+            body += (b'Content-Length: %d' % os.path.getsize(filename)) + crlf
+            logging.debug("encode body (without binary file) -%s, %s", type(body), body)
+            with open(filename, "rb") as fd:
                 body += crlf + fd.read() + crlf
-            body += '--' + boundary + '--' + crlf + crlf
-        return ('multipart/form-data; boundary=' + boundary, body)
+            body += b'--' + boundary + b'--' + crlf + crlf
+        return (b'multipart/form-data; boundary=' + boundary, body)
 
     def marshall(self, args, files, request):
         """
@@ -217,9 +246,12 @@ class DQMUpload(Executor):
         of the CGI script.
         """
         (contentType, body) = self.encode(args, files)
-        request.add_header('Content-Type', contentType)
-        request.add_header('Content-Length', str(len(body)))
-        request.add_data(body)
+        request.add_header(b'Content-Type', contentType)
+        if PY2:
+            request.add_header('Content-Length', str(len(body)))
+            request.add_data(body)
+        elif PY3:
+            request.data = body
         return
 
     def upload(self, url, args, filename):
@@ -240,11 +272,12 @@ class DQMUpload(Executor):
         logging.info(msg)
 
         handler = HTTPSAuthHandler(key=uploadProxy, cert=uploadProxy)
-        opener = urllib2.OpenerDirector()
+        opener = OpenerDirector()
         opener.add_handler(handler)
 
         # setup the request object
-        datareq = urllib2.Request(url + '/data/put')
+        url = decodeBytesToUnicode(url) if PY3 else encodeUnicodeToBytes(url)
+        datareq = Request(url + '/data/put')
         datareq.add_header('Accept-encoding', 'gzip')
         datareq.add_header('User-agent', ident)
         self.marshall(args, {'file': filename}, datareq)
@@ -252,7 +285,7 @@ class DQMUpload(Executor):
         if 'https://' in url:
             result = opener.open(datareq)
         else:
-            opener.add_handler(urllib2.ProxyHandler({}))
+            opener.add_handler(ProxyHandler({}))
             result = opener.open(datareq)
 
         data = result.read()

--- a/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py
+++ b/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py
@@ -1,0 +1,102 @@
+from __future__ import print_function
+import logging
+from future import standard_library
+standard_library.install_aliases()
+import urllib.request
+
+import unittest
+from nose.plugins.attrib import attr
+
+import os
+from functools import reduce
+from hashlib import md5
+
+from WMCore.WMSpec.Steps.Executors.DQMUpload import DQMUpload
+
+class StepPatch():
+    def __init__(self, upload=None):
+        self.upload = upload
+
+class UploadPatch():
+    def __init__(self):
+        self.proxy = False
+
+class DQMUpload_t(unittest.TestCase):
+    def setUp(self):
+        pass
+
+    def tearDown(self):
+        pass
+
+    @attr("integration")
+    def test_upload(self):
+        """
+        test_upload
+
+        This unittest has been used to validate the modernization of DQMUpload.
+        If everything is ok, this unittest is expected to pass.
+
+        However, this unittest sends a 144MB root file to the DQM. We convened 
+        that it is a bit too much for multiple daily checks, so we mark this
+        as "integration". See discussion at:
+        https://github.com/dmwm/WMCore/pull/9934#issuecomment-898855725
+        """
+        filename = "DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root"
+        filepath = os.path.join("/tmp", filename)
+        if not os.path.isfile(filepath):
+            urllib.request.urlretrieve(
+                "http://amaltaro.web.cern.ch/amaltaro/forAlan/%s" % filename, 
+                filepath)
+
+        dqm = DQMUpload()
+        upload_patch = UploadPatch()
+        step_patch = StepPatch(upload_patch)
+        dqm.step = step_patch
+
+        args = {}
+        # Preparing a checksum
+        blockSize = 0x10000
+        def upd(m, data):
+            m.update(data)
+            return m
+        with open(filepath, 'rb') as fd:
+            contents = iter(lambda: fd.read(blockSize), b'')
+            m = reduce(upd, contents, md5())
+        args['checksum'] = 'md5:%s' % m.hexdigest()
+        args['size'] = os.path.getsize(filepath)
+        headers, data = dqm.upload(
+            "https://cmsweb-testbed.cern.ch/dqm/dev/", 
+            args,
+            filepath)
+
+        logging.debug("headers: %s %s", type(headers), headers)
+        logging.debug("data: %s %s", type(data), data)
+
+        # Dqm-Status-Code == 100 is considered good
+        # Dqm-Status-Code == 300 was related to " File name does not match the expected convention"
+        self.assertEqual(headers.get("Dqm-Status-Code", None), '100')
+
+    @attr("integration")
+    def test_httppost(self):
+        """
+        test_httppost
+
+        This unittest has been used to validate the modernization of DQMUpload.
+        If everything is ok, this unittest is expected to pass.
+        """
+        filename = "DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root"
+        filepath = os.path.join("/tmp", filename)
+        if not os.path.isfile(filepath):
+            urllib.request.urlretrieve(
+                "http://amaltaro.web.cern.ch/amaltaro/forAlan/%s" % filename, 
+                filepath)
+
+        dqm = DQMUpload()
+        upload_patch = UploadPatch()
+        step_patch = StepPatch(upload_patch)
+        dqm.step = step_patch
+        dqm.step.upload.URL = "https://cmsweb-testbed.cern.ch/dqm/dev/"
+        dqm.httpPost(filepath)
+
+if __name__ == '__main__':
+    unittest.main()

